src/CameraOnvif.cpp

```diff
@@ -0,0 +1,408 @@
+#include "CameraOnvif.hpp"
+
+#include <gsoap/plugin/wsseapi.h>
+#include <gsoap/plugin/wsddapi.h>
+#include "onvif/soapDeviceBindingProxy.h"
+#include "onvif/soapImagingBindingProxy.h"
+#include "onvif/soapMediaBindingProxy.h"
+#include "onvif/soapPTZBindingProxy.h"
+#include "onvif/soapPullPointSubscriptionBindingProxy.h"
+#include "onvif/soapRemoteDiscoveryBindingProxy.h"
+#include "wsdd.nsmap"
+
+#include <iostream>
+
+using namespace std;
+using namespace camera_onvif;
+
+struct CameraOnvif::Private {
+    struct soap *soap = soap_new1(SOAP_XML_STRICT | SOAP_XML_CANONICAL | SOAP_C_UTFSTRING);
+    MediaBindingProxy *proxy_media;
+    ImagingBindingProxy *proxy_image;
+    _trt__GetVideoEncoderConfigurationOptionsResponse *video_options;
+}defined;
+
+CameraOnvif::CameraOnvif(/* args */) {
+    user = "admin";
+    pass = "camera01";
+    uri = "http://10.20.0.188/onvif/device_service";
+
+    this->init();
+}
+
+CameraOnvif::CameraOnvif(string user, string pass, string ip){
+    this->user = user;
+    this->pass = pass;
+    this->uri = "http://" + ip + "/onvif/device_service";
+
+    this->init();
+}
+
+CameraOnvif::CameraOnvif(string user, string pass, string ip, int width, int height){
```

Comment:
I don't think this constructor makes sense. The user will call setResolution after initialization if he wants to.

Reply:
Can be too, but the reason to do that is the problem with changing resolution to gstreamer, if I change the resolution with pipeline playing I will break it. So, making this I were forcing to change it in constructor and hiding it from user (I didn't put it as private but could be).

Reply:
For the record since we discussed it in person: you're constraining/designing the API of CameraOnvif because of limitations of something that is going to _use_ it. This is not a limitation of CameraOnvif itself, it's not an argument for the design of CameraOnvif.